tests/upgrades/test_performance_tuning.py

```diff
@@ -0,0 +1,153 @@
+                            "\nmongodb::server::storage_engine: 'wiredTiger'"
+
+
+class ScenarioPerfornaceTuning(TestCase):
+    """The test class contains pre-upgrade and post-upgrade scenarios to test
+    Performance Tuning utility
+
+    Test Steps:
+
+    1. Before Satellite upgrade, we will apply the medium tune size.
+    2. Once the size get apply we will wait for post upgrade test case to check
+    whether the size is same or not.
+    3. In post upgrade we will check whether satellite installer would able to change
+    the size or not
+
+    """
+
+    @pre_upgrade
+    def test_pre_performance_tuning_apply(self):
+        """In preupgrade scenario we will apply the medium tuning size.
+
+        :id: preupgrade-83404326-20b7-11ea-a370-48f17f1fc2e1
+
+        :steps:
+            1. Run satellite-installer --tuning medium --disable-system-checks.
+
+        :expectedresults: Medium tuning parameter should be applied.
+
+         """
+        with open('custom-hiera.yaml', 'w') as fd:
+            fd.write("{}\n".format(DEFAULT_CUSTOM_HIERA_DATA))
+            fd.write("{}".format(MEDIUM_TUNING_DATA))
+        ssh.upload_file('custom-hiera.yaml', '/etc/foreman-installer')
+        command_output = [data.strip(" ") for data in
+                          ssh.command('satellite-installer --disable-system-checks')
+                              .stdout]
+        assert "Success!" in command_output
+
+    @post_upgrade
+    def test_pre_performance_tuning_apply(self):
+        """In postupgrade scenario we will check the tune paramater state after upgrade
+
+        :id: postupgrade-31e26b08-2157-11ea-9223-001a4a1601d8
+
+        :steps:
+            1. Collect the current tuning state from satellite.yaml file.
+            2. Compare it with the expected value.
+            3. If Expected value get matched then this sceanrio would be passed.
+            1. Run satellite-installer --tuning medium --disable-system-checks.
+
+        :expectedresults: Medium tuning parameter should be applied.
+
+         """
+        tuning_state_after_upgrade = ssh.command('less /etc/foreman-installer/'
+                                                 'scenarios.d/satellite.yaml|'
+                                                 'grep "tuning: medium";if [ $? -eq 0 ]; '
+                                                 'then echo "PASSED"; else echo "FAILED";'
+                                                 ' fi').stdout
+        assert "PASSED" in tuning_state_after_upgrade
+        with open('custom-hiera.yaml', 'w') as fd:
+            fd.write("{}".format(DEFAULT_CUSTOM_HIERA_DATA))
+
+        ssh.upload_file('custom-hiera.yaml', '/etc/foreman-installer')
+        command_output = [data.strip(" ") for data in
+                          ssh.command('satellite-installer --tuning default '
+                                      '--disable-system-checks').stdout]
+        assert "Success!" in command_output
```

Comment:
I assume the goal here is to revert Satellite to default state after test succeeds.

I think it would be better idea to download file as-is in pre-upgrade and upload it in post-upgrade. Right now you hardcode current defaults in test. If defaults change, you will need to update this test (or, what is more likely, we will not notice, not update and post-upgrade will restore slightly different state than it was before pre-upgrade).

Also, you possibly want this in `try ... finally` block, to ensure it runs even if test itself failed for whatever reason.

Reply:
> I assume the goal here is to revert Satellite to default state after test succeeds.
> 
> I think it would be better idea to download file as-is in pre-upgrade and upload it in post-upgrade. Right now you hardcode current defaults in test. If defaults change, you will need to update this test (or, what is more likely, we will not notice, not update and post-upgrade will restore slightly different state than it was before pre-upgrade).
> 
> Also, you possibly want this in `try ... finally` block, to ensure it runs even if test itself failed for whatever reason.

Thanks, Mirek, The test cases would be moved from one version to another, In this case, we will need to track the changes of the custom-hiera.yaml file, that's the reason I have written two test cases to verify the custom-hiera.yaml file in pre-upgrade and post-upgrade. If something gets change after an upgrade then we should aware of it.

Yes, you are right `try ... finally` block is required to remediate if any failure happens with default tune parameters after post-upgrade.